src/eth1spec/base_types.py

```diff
@@ -0,0 +1,553 @@
+                raise ValueError()
+
+        if not isinstance(left, int):
+            return NotImplemented
+
+        if left < 0:
+            raise ValueError()
+
+        result = super(Uint, self).__rpow__(left, modulo)
+        return self.__class__(result)
+
+    def __ipow__(self, right: int, modulo: Optional[int] = None) -> "Uint":
+        return self.__pow__(right, modulo)
+
+    # TODO: Implement and, or, xor, neg, pos, abs, invert, ...
+
+    def to_be_bytes32(self) -> "Bytes32":
+        """
+        Converts this arbitrarily sized unsigned integer into its big endian
+        representation with exactly 32 bytes.
+
+        Returns
+        -------
+        big_endian : `Bytes32`
+            Big endian (most significant bits first) representation.
+        """
+        return self.to_bytes(32, "big")
+
+    def to_be_bytes(self) -> "Bytes":
+        """
+        Converts this arbitrarily sized unsigned integer into its big endian
+        representation.
+
+        Returns
+        -------
+        big_endian : `Bytes`
+            Big endian (most significant bits first) representation.
+        """
+        bit_length = self.bit_length()
+        byte_length = (bit_length + 7) // 8
+        return self.to_bytes(byte_length, "big")
+
+
+class U256(int):
+    """
+    Unsigned positive integer, which can represent `0` to `2 ** 256 - 1`,
+    inclusive.
+    """
+
+    MAX_VALUE: "U256"
+
+    __slots__ = ()
+
+    @classmethod
+    def from_be_bytes(cls: Type, buffer: "Bytes") -> "U256":
+        """
+        Converts a sequence of bytes into an arbitrarily sized unsigned integer
+        from its big endian representation.
+
+        Parameters
+        ----------
+        buffer : `Bytes`
+            Bytes to decode.
+
+        Returns
+        -------
+        self : `U256`
+            Unsigned integer decoded from `buffer`.
+        """
+        if len(buffer) > 32:
+            raise ValueError()
+
+        return cls(int.from_bytes(buffer, "big"))
+
+    def __new__(cls: Type, value: int) -> "U256":
+        if not isinstance(value, int):
+            raise TypeError()
+
+        if value < 0 or value >= 2 ** 256:
+            raise ValueError()
+
+        return super(cls, cls).__new__(cls, value)
+
+    def __radd__(self, left: int) -> "U256":
+        return self.__add__(left)
+
+    def __add__(self, right: int) -> "U256":
+        result = self.unchecked_add(right)
+
+        if result == NotImplemented:
+            return NotImplemented
+
+        return self.__class__(result)
+
+    def unchecked_add(self, right: int) -> int:
+        if not isinstance(right, int):
+            return NotImplemented
+
+        if right < 0 or right > self.MAX_VALUE:
+            raise ValueError()
+
+        return super(U256, self).__add__(right)
+
+    def wrapping_add(self, right: int) -> "U256":
+        result = self.unchecked_add(right)
+
+        if result == NotImplemented:
+            return NotImplemented
```

Comment:
Returning this exception type rather than raising it is confusing to me.  Any specific reason for this?  If it was simply raised at the points where things aren't implemented then you would get clean stacktraces showing you where these are coming from.  As it stands now it will be hard to determine the origin of these errors.

Reply:
`NotImplemented` isn't an exception (see https://docs.python.org/3/library/constants.html#NotImplemented).

This is just how these functions are supposed to work, apparently.

Reply:
I was reading these as: https://docs.python.org/3/library/exceptions.html#NotImplementedError